File: src/openagents/tool/shell_tool.py
import asyncio
import logging
import subprocess
import sys

import anyio

POWERSHELL_UTF8_SETTING = """# UTF-8 encoding setting
[Console]::OutputEncoding = [System.Text.UTF8Encoding]::new()
$OutputEncoding = [System.Text.UTF8Encoding]::new()

# Basic file operation commands UTF-8 encoding
$PSDefaultParameterValues['Out-File:Encoding'] = 'utf8'
$PSDefaultParameterValues['Set-Content:Encoding'] = 'utf8'
$PSDefaultParameterValues['Add-Content:Encoding'] = 'utf8'
"""


async def setup_powershell_utf8():
    # Windows才配置
    if not sys.platform.startswith("win"):
        return
    # 配置 Windows PowerShell 5.1 输出编码
    home_dir = await anyio.Path.home()
    profile_path = home_dir / "Documents" / "WindowsPowerShell" / "Microsoft.PowerShell_profile.ps1"
    await profile_path.parent.mkdir(parents=True, exist_ok=True)
    await profile_path.write_text(POWERSHELL_UTF8_SETTING, encoding="utf-8")
    # 配置 Windows PowerShell 7 输出编码
    profile_path = home_dir / "Documents" / "PowerShell" / "Microsoft.PowerShell_profile.ps1"
    await profile_path.parent.mkdir(parents=True, exist_ok=True)
    await profile_path.write_text(POWERSHELL_UTF8_SETTING, encoding="utf-8")
    # 输出日志
    logging.info("PowerShell default encoding configured")


async def execute(cmd_and_args: list[str], work_dir: str) -> tuple[str, bool]:
    process = await asyncio.create_subprocess_exec(*cmd_and_args, cwd=work_dir, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
    stdout, stderr = await process.communicate()
    tool_content = f"{stdout.decode('utf-8', errors='replace')}{stderr.decode('utf-8', errors='replace')}"
    is_error = process.returncode != 0
    return tool_content, is_error

File: src/service/tool/shell_tool.rs
// Shell 执行工具
use tokio::process::Command;

use super::ToolResult;

// PowerShell UTF-8 配置内容
const POWERSHELL_UTF8_SETTING: &str = r#"# UTF-8 encoding setting
[Console]::OutputEncoding = [System.Text.UTF8Encoding]::new()
$OutputEncoding = [System.Text.UTF8Encoding]::new()

# Basic file operation commands UTF-8 encoding
$PSDefaultParameterValues['Out-File:Encoding'] = 'utf8'
$PSDefaultParameterValues['Set-Content:Encoding'] = 'utf8'
$PSDefaultParameterValues['Add-Content:Encoding'] = 'utf8'
"#;

// Windows 下配置 PowerShell UTF-8 编码
pub async fn setup_powershell_utf8() {
    if !cfg!(windows) {
        return;
    }

    let home_dir = crate::config::home_dir();
    // 依次配置 Windows PowerShell 5.1 与 PowerShell 7 的输出编码
    for profile_dir in ["WindowsPowerShell", "PowerShell"] {
        let profile_path = std::path::Path::new(&home_dir)
            .join("Documents")
            .join(profile_dir)
            .join("Microsoft.PowerShell_profile.ps1");
        if let Some(parent) = profile_path.parent() {
            let _ = tokio::fs::create_dir_all(parent).await;
        }
        let _ = tokio::fs::write(&profile_path, POWERSHELL_UTF8_SETTING).await;
    }

    tracing::info!("PowerShell default encoding configured");
}

// 执行 shell 命令
pub async fn execute(cmd_and_args: &[String], work_dir: &str) -> ToolResult {
    if cmd_and_args.is_empty() {
        return ("No command provided".to_string(), true);
    }

    let mut cmd = Command::new(&cmd_and_args[0]);
    for arg in &cmd_and_args[1..] {
        cmd.arg(arg);
    }
    cmd.current_dir(work_dir);
    cmd.stdout(std::process::Stdio::piped());
    cmd.stderr(std::process::Stdio::piped());
    // Windows 下隐藏子进程控制台窗口, 避免桌面模式调用外部模型时弹出黑框
    #[cfg(windows)]
    {
        cmd.creation_flags(0x08000000);
    }

    match cmd.output().await {
        Ok(output) => {
            let stdout = String::from_utf8_lossy(&output.stdout);
            let stderr = String::from_utf8_lossy(&output.stderr);
            let tool_content = format!("{}{}", stdout, stderr);
            let is_error = !output.status.success();
            (tool_content, is_error)
        }
        Err(e) => (format!("Failed to execute command: {}", e), true),
    }
}
